// @angular/platform-browser/esm/src/animate/browser_details.js
import { Injectable } from '@angular/core';
import { Math } from '../../src/facade/math';
import { getDOM } from '../dom/dom_adapter';
export class BrowserDetails {
    constructor() {
        this.elapsedTimeIncludesDelay = false;
        this.doesElapsedTimeIncludesDelay();
    }
    /**
     * Determines if `event.elapsedTime` includes transition delay in the current browser.  At this
     * time, Chrome and Opera seem to be the only browsers that include this.
     */
    doesElapsedTimeIncludesDelay() {
        var div = getDOM().createElement('div');
        getDOM().setAttribute(div, 'style', `position: absolute; top: -9999px; left: -9999px; width: 1px;
      height: 1px; transition: all 1ms linear 1ms;`);
        // Firefox requires that we wait for 2 frames for some reason
        this.raf((timestamp) => {
            getDOM().on(div, 'transitionend', (event) => {
                var elapsed = Math.round(event.elapsedTime * 1000);
                this.elapsedTimeIncludesDelay = elapsed == 2;
                getDOM().remove(div);
            });
            getDOM().setStyle(div, 'width', '2px');
        }, 2);
    }
    raf(callback, frames = 1) {
        var queue = new RafQueue(callback, frames);
        return () => queue.cancel();
    }
}
BrowserDetails.decorators = [
    { type: Injectable },
];
BrowserDetails.ctorParameters = [];
class RafQueue {
    constructor(callback, frames) {
        this.callback = callback;
        this.frames = frames;
        this._raf();
    }
    _raf() {
        this.currentFrameId =
            getDOM().requestAnimationFrame((timestamp) => this._nextFrame(timestamp));
    }
    _nextFrame(timestamp) {
        this.frames--;
        if (this.frames > 0) {
            this._raf();
        }
        else {
            this.callback(timestamp);
        }
    }
    cancel() {
        getDOM().cancelAnimationFrame(this.currentFrameId);
        this.currentFrameId = null;
    }
}
